### ISSUE TITLE
After yesterday night automatic windows 10 update, ds4windows doesn't work anymore

### ISSUE THREAD

**Issue by blitzcloud** [Jays2Kings/DS4Windows#277]
Not in shared nor exclusive mode.

**Comment by Ryochan7** [comment]
Just installed the latest Windows update. Exclusive mode does not work automatically for me anymore. Exclusive mode does work in DS4Windows after manually disabling and then re-enabling the controller in the Device Manager. I am connecting my controller via USB.

**Comment by Ryochan7** [comment]
Oddly enough, automatically enabling exclusive mode works fine with the small changes made in my lag_reduce branch. I have tested this multiple times with that branch and with the vanilla version. The lag_reduce version works automatically for a new attached controller like DS4Windows did before the latest Windows update. The vanilla DS4Windows cannot automatically get exclusive mode access without a manual workaround or running the lag_reduce version first.

Edit: Spoke too soon. lag_reduce is not the solution. It really seems like running a debug version mattered more since a compiled debug version of jay works fine as well. It looks like the manual workaround will be required for now.

Edit 2: At this point, it looks like the option "Prefer 32-bit" for the Visual Studio project is the main culprit. When that option is enabled and an executable is built, exclusive mode does not work properly. If the option is disabled then exclusive mode works fine with that executable.

**Comment by Ryochan7** [comment]
The solution seems to be removing the option "Prefer 32-bit" from the project settings before building the executable. A new exe would have to be compiled and released with the change made to the project settings. Otherwise, the manual workaround of disabing and then re-enabling the device in Device Manager would be needed. Also, a separate 32-bit version and 64-bit version of the program would have to be released.

**Comment by Ryochan7** [comment]
As long as "Prefer 32-bit" is disabled in the project settings, the AnyCPU platform target can still be used. Here is an updated exe with the relevant change. This build is based on the current status of the jay branch. On my system, both USB and Bluetooth connections work and exclusive mode works for both.

https://www.dropbox.com/s/9q95o2fbp5cw0k3/DS4Windows_Test_20170316.zip

**Comment by lmhan** [comment]
Hey thanks the new exe fixed my problem, cheers!

**Comment by GingerAdonis** [comment]
I'm still having Bluetooth problems when using your binary @Ryochan7 (1.4.53 release or the build mentioned in this issue). I can't seem to properly sync/pair the controllers anymore. The controllers sometimes hang, or simply don't connect. I do see the controllers appear in the Bluetooth screen and the PC thinks pairing goes fine. Any idea with this could be?

(USB works fine)

ScpToolkit also stopped working for me 5-6 days ago.

**Comment by Ryochan7** [comment]
Even though it has been rare, I have encountered some weird behavior when testing Bluetooth. Sync issues occurred a couple of times and sometimes the controller would be marked as disabled and would have to be re-enabled in Device Manager before it could be used in DS4Windows. There hasn't been anything that I have found yet that would directly cause those problems.

Lately, I have only been concerned with decreasing input lag. The latest commits include changes to IO routines which might help with hang issues in DS4Windows along with decreasing input lag. I haven't had problems since the changes were completed. More testing will need to be done before a new build would be made.

**Comment by GingerAdonis** [comment]
Too bad. Was hoping to continue using my PS4 controllers.

It's great that you're picking up the project. :)

**Comment by Ryochan7** [comment]
Hopefully something can be resolved at some point. It looks like Bluetooth support can be a big problem. On my system, even though I can use a DS4 controller via Bluetooth in DS4Windows, it does not work in the latest stable version of InputMapper. Whenever I try to get InputMapper to use the controller via Bluetooth, the controller just disconnects. The controller is detected and connects when using a USB connection though.

**Comment by JohannesAtGit** [comment]
My problem is, that the Bluetooth connection is very unstable under Windows 10.
After a while the controller looses the connection and the blue light turns grey or out. In the log are many latency errors.
The distance between usb bluetooth adapter and controller is just 3m.

**Comment by GingerAdonis** [comment]
Just for my record. The controller connected via Bluetooth should have the general Bluetooth driver by Microsoft, right?

**Comment by Ryochan7** [comment]
Yes. The general driver would be used for the controller connection. It is too bad that right now there doesn't seem to be a good way to know when the process of enumerating Bluetooth controllers fails. I wouldn't be surprised if the HID device is initially detected and fails later in the process of establishing the connection. I doubt it would help but you can check out the latest release and see if you have better luck with that version.

https://github.com/Ryochan7/DS4Windows/releases/download/v1.4.54/DS4Windows_1.4.54_x64.zip

**Comment by GingerAdonis** [comment]
It doesn't fix my problem unfortunately. It connects and hangs at a light blue color. It's never detected by either Windows (apart from Bt connection) or DS4Windows/ScpToolkit. Thanks for trying helping me out anyways.

**Comment by GingerAdonis** [comment]
Installed a clean Windows 10 with latest updates on my laptop (with Bluetooth): Everything works fine.
Ran a portable Windows 10 without latest updates: Still Bluetooth problems. Updating Windows won't fix these problems.

I guess my adapter decided to break. There's no need to look into this anymore. Will get an Bt USB dongle and hope I'll get up and running again. :)

**Comment by GingerAdonis** [comment]
Can confirm: For me it was a sudden radio problem. New USB Bluetooth dongle works fine.
